### PR TITLE
Automatically undistribute citus local tables when no more fkeys with reference tables

### PR DESCRIPTION
**Pull request by onurctirtir**
As part of #4415, we need to undistribute citus local tables when they are not involved in any foreign key relationships with reference tables (directly or indirectly).
For that reason, whenever a table constraint is dropped, we check if there are such citus local tables and undistribute them if so.

### CODE REVIEW

File: src/backend/distributed/commands/utility_hook.c [hunk 1]
```diff
@@ -615,6 +661,140 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+	if (!EnableLocalReferenceForeignKeys)
+	{
+		/*
+		 * If foreign keys between reference tables and local tables are
+		 * disabled, then user might be using create_citus_local_table for
+		 * their own purposes. In that case, we should not undistribute
+		 * citus local tables. We also disable this flag in some regression
+		 * tests for unit testing of citus local tables.
+		 */
+		return false;
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:749]: maybe error out in `create_citus_local_table` udf when this guc is on

File: src/backend/distributed/utils/foreign_key_relationship.c [hunk 2]
```diff
@@ -144,6 +145,30 @@ GetForeignKeyConnectedRelationIdList(Oid relationId)
...
 
 
+/*
+ * ConnectedToReferenceTableViaFKey returns true if given relationId is
+ * connected to a reference table via its foreign key subgraph.
+ */
+bool
+ConnectedToReferenceTableViaFKey(Oid relationId)
+{
+	/* we can't rely on foreign key graph */
```
Comment (onurctirtir) [inline src/backend/distributed/utils/foreign_key_relationship.c:155]: better comment

File: src/backend/distributed/commands/utility_hook.c (hunk 1 above)
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:749]: maybe error out in create_citus_local_table udf when this guc is on

File: src/backend/distributed/commands/utility_hook.c [hunk 3]
```diff
@@ -165,6 +176,40 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	utilityHookLevel--;
+}
+
+
+static void
+citus_ProcessUtility(PlannedStmt *pstmt,
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:202]: comment and better name, maybe `multi_ProcessUtility_Internal`

File: src/backend/distributed/commands/drop_distributed_table.c [hunk 4]
```diff
@@ -148,3 +150,21 @@ MasterRemoveDistributedTableMetadataFromWorkers(Oid relationId, char *schemaName
...
+ * to set ConstraintDropped to true.
+ * This udf is designed to be called from citus_drop_trigger to tell us we
+ * dropped a table constraint.
+ */
+Datum
+notify_constraint_dropped(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* TODO: maybe error out if not called via event trigger ? */
```
Comment (onurctirtir) [inline src/backend/distributed/commands/drop_distributed_table.c:166]: .

File: src/backend/distributed/commands/utility_hook.c [hunk 5]
```diff
@@ -165,6 +177,38 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 		return;
 	}
 
+	PG_TRY();
+	{
+		utilityHookLevel++;
+
+		bool executedDDLJob = citus_ProcessUtility(pstmt, queryString, context,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:184]: we already have another function named `CitusProcessUtility`. We should come up with a better name for either of these

File: src/backend/distributed/utils/foreign_key_relationship.c [hunk 6]
```diff
@@ -144,6 +145,30 @@ GetForeignKeyConnectedRelationIdList(Oid relationId)
...
+	InvalidateForeignKeyGraph();
+
+	List *fkeyConnectedRelations = GetForeignKeyConnectedRelationIdList(relationId);
```
Comment (onderkalaci) [inline src/backend/distributed/utils/foreign_key_relationship.c:158]: We somewhere have a function which does this loop?

File: src/test/regress/expected/alter_table_set_access_method.out [hunk 7]
```diff
@@ -356,6 +356,7 @@ SELECT create_reference_table('table_type_ref');
 (1 row)
 
 CREATE TABLE table_type_citus_local(a INT);
+SET citus.enable_local_reference_table_foreign_keys TO OFF;
```
Comment (onderkalaci) [inline src/test/regress/expected/alter_table_set_access_method.out:359]: I guess we should remove these `SET citus.enable_local_reference_table_foreign_keys TO OFF;` calls so that we can check the CI easier

File: src/backend/distributed/commands/utility_hook.c (hunk 5 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:184]: yeah thinking about the same thing
https://github.com/citusdata/citus/pull/4538#discussion_r560016567

File: src/backend/distributed/utils/foreign_key_relationship.c (hunk 6 above)
Comment (onurctirtir) [reply src/backend/distributed/utils/foreign_key_relationship.c:158]: we have a loop in `CascadeOperationForConnectedRelations` but it's slightly different than this one

**Comment by onurctirtir** [comment]
With last commit, now regression test differences are much more reasonable and mostly good, now I can sleep at peace :)

File: src/backend/distributed/commands/utility_hook.c (hunk 1 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:789]: (no, we won't do that)

File: src/backend/distributed/commands/utility_hook.c (hunk 3 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: one day we should remove such `multi` word from function & file names

File: src/backend/distributed/commands/utility_hook.c [hunk 8]
```diff
@@ -612,6 +658,180 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+	}
+
+	if (IsDropSchemaOrTable(parsetree))
+	{
+		/*
+		 * TODO: We could specifically check if it is a:
+		 * - DROP TABLE statement that drops a reference or a citus local table
+		 *   that has a foreign key relationship with a postgres table, or a
+		 * - DROP SCHEMA statement that again drops such a citus table
+		 * but this is more easy and the other approach is much more prune to
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:674]: >the other approach is much more prune to
		 * deadlocks. 
		 
 Why is so?

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:674]: For example when dropping a schema, we need to check every citus table in that schema and I think we need to acquire some locks on those citus tables when doing that.
But actually that drop schema/table command could also be a false alarm, so even when we are using local execution, possibility of ending-up with a deadlock would be higher I think.

What do you think ?

File: src/backend/distributed/commands/alter_table.c [hunk 9]
```diff
@@ -479,6 +480,8 @@ AlterTableSetAccessMethod(TableConversionParameters *params)
 TableConversionReturn *
 ConvertTable(TableConversionState *con)
 {
+	SetLocalExecutionStatus(LOCAL_EXECUTION_REQUIRED);
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/alter_table.c:483]: yeah, like in https://github.com/citusdata/citus/pull/4540, we should also make sure that we are using local execution. For example drop trigger might drop something via remote execution and that might end up undistributing citus local tables in remote session

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 10]
```diff
@@ -366,7 +365,7 @@ ExecuteCascadeOperationForRelationIdList(List *relationIdList,
 		{
 			case CASCADE_FKEY_UNDISTRIBUTE_TABLE:
 			{
-				if (IsCitusTable(relationId))
+				if (RelationExists(relationId) && IsCitusTable(relationId))
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:368]: undistributing one of the those tables in loop might automatically undistribute another citus local table (due to drop fkey command), we might hit to a citus local table that is already undistrbiuted

File: src/test/regress/expected/ref_citus_local_fkeys.out [hunk 11]
```diff
@@ -160,25 +183,27 @@ ERROR:  cannot define foreign key constraint, foreign keys from reference tables
...
+NOTICE:  executing the command locally: SELECT worker_apply_inter_shard_ddl_command (1506007, 'ref_citus_local_fkeys', 1506006, 'ref_citus_local_fkeys', 'ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES citus_local_table(l1) ON DELETE NO ACTION;')
 -- show that adding/dropping foreign keys from reference to citus local
 -- tables works fine with remote execution too
 SET citus.enable_local_execution TO OFF;
 ALTER TABLE reference_table DROP CONSTRAINT fkey_ref_to_local;
+ERROR:  cannot switch local execution status from local execution disabled to local execution enabled since it can cause visibility problems in the current transaction
 ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_local FOREIGN KEY(r1) REFERENCES citus_local_table(l1) ON DELETE NO ACTION;
+ERROR:  constraint "fkey_ref_to_local" for relation "reference_table" already exists
 SET citus.enable_local_execution TO ON;
```
Comment (onurctirtir) [inline src/test/regress/expected/ref_citus_local_fkeys.out:202]: only suspicious difference was this, it would be resolved after we merge https://github.com/citusdata/citus/pull/4540

File: src/backend/distributed/commands/alter_table.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/alter_table.c:483]: So, lets add a clarifying comment

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:368]: It means that the graph or the relationList is out-dated. We should be invalidating the graph / re-creating the tables to cascade?

File: src/backend/distributed/commands/utility_hook.c (hunk 8 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:674]: The session that is executing this piece of the code already acquired very heavy locks on the tables. I don't see any additional potential deadlocks.

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:368]: >It means that the graph or the relationList is out-dated

True, but we already act on relations that graph returned. While iterating on those relations, we need to re-enter to loop again and again, not sure how we can implement that

File: src/backend/distributed/commands/utility_hook.c (hunk 3 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: we will not make this change in this pr

File: src/backend/distributed/utils/foreign_key_relationship.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/foreign_key_relationship.c:158]: I meant we can use `RelationIdListHasReferenceTable(fkeyConnectedRelations)`

File: src/backend/distributed/commands/utility_hook.c [hunk 12]
```diff
@@ -612,6 +658,180 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+		 * deadlocks.
+		 */
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:676]: We also need to handle `DROP OWNED BY ..` similar to `IsDropSchemaOrTable `

File: src/backend/distributed/commands/utility_hook.c [hunk 13]
```diff
@@ -644,6 +666,180 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+
+	if (UtilityHookLevel != 1)
+	{
+		/* we only perform this operation for top level DDL command */
+		return false;
+	}
+
+	ResetConstraintDropped();
+
+	if (!EnableLocalReferenceForeignKeys)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:788]: it is better to use `ShouldEnableLocalReferenceForeignKeys`

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: 1c7df53f499ed6c6cd81a3207f9269eca565f9f5

File: src/backend/distributed/commands/alter_table.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/alter_table.c:491]: as we talked in person, this should go to undistribute_table, not `ConvertTable `

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:368]: I'd like to debug the case where we hit this case. But, you can leave as-is for now that's OK.

File: src/backend/distributed/commands/utility_hook.c [hunk 14]
```diff
@@ -644,6 +666,179 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+			continue;
+		}
+
+		/*
+		 * Citus local table is not connected to any reference table, then
+		 * undistribute it via cascade.
+		 */
+		TableConversionParameters params = {
+			.relationId = citusLocalTableId,
+			.cascadeViaForeignKeys = true
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:756]: This logic seems slightly off. We already get all Citus local tables, and still use `cascadeViaForeignKeys = true`. Could this be the reason that you see `relation not exists ` errors?


(Our initial idea was that we would be able to find the exact relation which is affected by the drop, that's why we ended up `cascadeViaForeignKeys `. With the current approach, we shouldn't need this?)

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:756]: we might still have citus local tables connected to each other but not to a reference table, so I guess we need to `cascadeViaForeignKeys `

Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:756]: Well, those should be qualified to be dropped individually as `ConnectedToReferenceTableViaFKey` will return false for all of them?

File: src/backend/distributed/commands/utility_hook.c [hunk 15]
```diff
@@ -237,8 +244,23 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 		return;
 	}
 
-	ProcessUtilityInternal(pstmt, queryString, context,
-						   params, queryEnv, dest, completionTag);
+	UtilityHookLevel++;
+
+	PG_TRY();
+	{
+		bool executedDDLJob = ProcessUtilityInternal(pstmt, queryString, context,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:251]: maybe call this `executedOnCitusTable` or such to make it a little more clear

File: src/backend/distributed/commands/utility_hook.c [hunk 16]
```diff
@@ -644,6 +666,179 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 		 */
 		CitusHasBeenLoaded();
 	}
+
+	if (list_length(ddlJobs) != 0)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:670]: can we move this logic into a function and maybe call it `ExecutedOnCitusTable `

File: src/backend/distributed/commands/alter_table.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/alter_table.c]: I did that in 393664b, but now we have some deadlock issues :/

File: src/backend/distributed/commands/utility_hook.c (hunk 16 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c]: Or, DDLOnCitusTable

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: makes sense, I will address this

File: src/backend/distributed/commands/utility_hook.c [hunk 17]
```diff
@@ -165,6 +178,38 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
 	}
 
+	PG_TRY();
+	{
+		utilityHookLevel++;
+
+		bool executedDDLJob = citus_ProcessUtility(pstmt, queryString, context,
+												   params, queryEnv, dest,
+												   completionTag);
+		UndistributeCitusLocalTablesIfNeeded(executedDDLJob);
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: In 8304956, actually addressed this review.
However then I realized that we should actually reset `ConstraintDropped` if we did not execute a ddl on a citus table.
So moved `executedDDLJob` check back into `UndistributeCitusLocalTablesIfNeeded->ShouldUndistributeCitusLocalTables` in 1f6bc14

File: src/backend/distributed/commands/utility_hook.c (hunk 14 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:811]: yes but we will still need to frop foreign key constraints on each of such citus local tables, then it would nearly be copy/paste of cascade via foreign keys logic

File: src/backend/distributed/commands/alter_table.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/alter_table.c]: reverted this change in the end

File: src/backend/distributed/utils/foreign_key_relationship.c (hunk 6 above)
Comment (onurctirtir) [reply src/backend/distributed/utils/foreign_key_relationship.c:158]: true, I will address  that

File: src/backend/distributed/commands/drop_distributed_table.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/drop_distributed_table.c:166]: ping on this

File: src/backend/distributed/commands/utility_hook.c [hunk 18]
```diff
@@ -237,8 +250,26 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
-	ProcessUtilityInternal(pstmt, queryString, context,
-						   params, queryEnv, dest, completionTag);
+	UtilityHookLevel++;
+
+	PG_TRY();
+	{
+		bool ddlExecutedOnCitusTable = ProcessUtilityInternal(pstmt, queryString, context,
+															  params, queryEnv, dest,
+															  completionTag);
+		if (UtilityHookLevel == 1)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:260]: some ideas to improve the flow a bit:
- passing the `ddlExecutedOnCitusTable ` flag down is hard to follow 
- Makes the calls to `ResetConstraintDropped ` way deep down and too implicit/error prone
- We should reset in the pg_catch as well because the transaction would rollback and the flag might be set to true



Please see https://github.com/citusdata/citus/commit/e4a3d97002d2a93f0282542c119460969e323632 for suggestions

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:260]: I will address that too

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:389]: All the tests pass even if I remove this check `RelationExists `.  Can you share steps that require this check?

https://github.com/citusdata/citus/commit/410b9b7ab505a14e41ec5d400df594853f17ba8a

File: src/backend/distributed/commands/utility_hook.c (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: true, addressed this at 5bae03d

File: src/backend/distributed/commands/drop_distributed_table.c (hunk 4 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/drop_distributed_table.c:166]: (looking for how we can understrand if we are in drop trigger)

Comment (onderkalaci) [reply src/backend/distributed/commands/drop_distributed_table.c:166]: check for `f (!CALLED_AS_TRIGGER(fcinfo))` in the code?

Comment (onderkalaci) [reply src/backend/distributed/commands/drop_distributed_table.c:166]: But I guess we may not be able to do it in this function

File: src/backend/distributed/commands/utility_hook.c [hunk 19]
```diff
@@ -647,6 +691,134 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+ * subgraphs.
+ */
+static void
+UndistributeDisconnectedCitusLocalTables(void)
+{
+	List *citusLocalTableIdList = CitusTableTypeIdList(CITUS_LOCAL_TABLE);
+	citusLocalTableIdList = SortList(citusLocalTableIdList, CompareOids);
+
+	/* acquire ExclusiveLock as UndistributeTable does */
+	LockRelationsWithLockMode(citusLocalTableIdList, ExclusiveLock);
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:706]: why do we need such an aggressive lock if we're just doing a few checks?

Maybe RowExclusiveLock to conflict with foreign key creation. Also seems like we could release locks if nothing happens / just keep whatever locks UndistributeTable takes.

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:389]: it was required before doing SetLocalLocal thing, now it's not needed, I will remove this check

File: src/backend/distributed/commands/utility_hook.c (hunk 19 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:706]: both makes sense, (iirc, `ShareRowExclusiveLock` is the lock that we should acquire for concurrent fkey creation)

File: src/backend/distributed/commands/utility_hook.c [hunk 20]
```diff
@@ -88,6 +96,8 @@ static void IncrementUtilityHookCountersIfNecessary(Node *parsetree);
 static void PostStandardProcessUtility(Node *parsetree);
 static void DecrementUtilityHookCountersIfNecessary(Node *parsetree);
 static bool IsDropSchemaOrDB(Node *parsetree);
+static void UndistributeDisconnectedCitusLocalTables(void);
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:99]: 1) implement a udf for that static function
2) error out for create_citus_local_table when `ShouldEnableForeignKeys` returns false

File: src/backend/distributed/commands/drop_distributed_table.c (hunk 4 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/drop_distributed_table.c]: afaik, this `CALLED_AS_TRIGGER` actually checks if function itself is called as trigger.
But with that (7507d32) commit, we shouldn't even need to error out ?

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c]: f5e905bb4163554dd2ad814000e06516cb2e09f1

File: src/backend/distributed/commands/utility_hook.c (hunk 14 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:737]: > Yes but we will still need to drop foreign key constraints on each of such citus local tables,

Ok, this should be in the comment. Plus, do we have a test covers this scenario? I think we should

File: src/backend/distributed/metadata/metadata_cache.c [hunk 21]
```diff
@@ -3797,6 +3797,27 @@ InvalidateMetadataSystemCache(void)
 }
 
 
+/*
+ * SchemaHasAnyReferenceOrCitusLocalTables returns true if given schema has
+ * any reference tables or citus local tables.
+ */
+bool
+SchemaHasAnyReferenceOrCitusLocalTables(Oid schemaId)
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_cache.c:3805]: we don't use this anymore

File: src/test/regress/sql/auto_undist_citus_local.sql [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+    res INT := 0;
+begin
+		PERFORM 1/0;
+    	EXCEPTION
+             when others then
+        CALL drop_constraint_via_proc();
+    commit;
+end;$$;
+
+CALL drop_constraint_via_proc_exception();
```
Comment (onderkalaci) [inline src/test/regress/sql/auto_undist_citus_local.sql:257]: Needs few more basic tests:
- A regular plpgsql function (e.g., use `SELECT func()`) which does something similar
- DROP OWNED BY
- Drop multiple tables 
- Drop in a TX block
- DROP in a save point, and rollback to pre-drop case

File: src/backend/distributed/commands/utility_hook.c [hunk 23]
```diff
@@ -647,6 +693,134 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+		return false;
+	}
+
+	if (!IsCoordinator())
+	{
+		/* we should not perform this operation in worker nodes */
+		return false;
+	}
+
+	if (!InCoordinatedTransaction())
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:787]: should we move these two above `ShouldEnableLocalReferenceForeignKeys `? These two seem cheaper to call?

File: src/backend/distributed/metadata/metadata_cache.c (hunk 21 above)
Comment (onurctirtir) [reply src/backend/distributed/metadata/metadata_cache.c]: d2feade

File: src/backend/distributed/commands/utility_hook.c (hunk 18 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:259]: I guess I addressed them in 592b817 & 2107c7d

File: src/backend/distributed/commands/utility_hook.c (hunk 23 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: makes sense e7644b8

File: src/backend/distributed/commands/utility_hook.c (hunk 14 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:734]: yeah I will add such tests

File: src/backend/distributed/commands/utility_hook.c (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: anyway removed that code, so resolving

File: src/backend/distributed/commands/utility_hook.c (hunk 16 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: not relying on ddljobs now, so resolving

File: src/backend/distributed/commands/utility_hook.c (hunk 15 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: removed that variable, so resolving

File: src/backend/distributed/commands/utility_hook.c (hunk 12 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: now we don't need to handle such commands separately but I need to add tests as mentioned in https://github.com/citusdata/citus/pull/4538#discussion_r561928566, so resolving this one

File: src/backend/distributed/commands/utility_hook.c [hunk 24]
```diff
@@ -647,6 +688,136 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+	{
+		/* not interacting with any Citus objects */
+		return false;
+	}
+
+	if (application_name != NULL && strcmp(application_name, CITUS_APPLICATION_NAME) == 0)
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:773]: might be nice to have a utility function that we can then also use in connection_management.c

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql [hunk 25]
```diff
@@ -1,9 +1,15 @@
+CREATE OR REPLACE FUNCTION notify_constraint_dropped()
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql:1]: should probably have a pg_catalog prefix

File: src/backend/distributed/commands/utility_hook.c (hunk 24 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: makes sense 008be03

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql (hunk 25 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql]: addressed 09e8170

File: src/backend/distributed/commands/alter_table.c [hunk 26]
```diff
@@ -487,6 +488,9 @@ AlterTableSetAccessMethod(TableConversionParameters *params)
 TableConversionReturn *
 ConvertTable(TableConversionState *con)
 {
+	bool oldEnableLocalReferenceForeignKeys = EnableLocalReferenceForeignKeys;
+	SetLocalEnableLocalReferenceForeignKeys(false);
```
Comment (marcocitus) [inline src/backend/distributed/commands/alter_table.c:492]: Could use a comment on why we need to do this.

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 27]
```diff
@@ -282,8 +282,28 @@ DropRelationIdListForeignKeys(List *relationIdList, int fKeyFlags)
 void
 DropRelationForeignKeys(Oid relationId, int fKeyFlags)
 {
+	bool oldEnableLocalReferenceForeignKeys = EnableLocalReferenceForeignKeys;
```
Comment (marcocitus) [inline src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:285]: Could use a comment

File: src/backend/distributed/commands/alter_table.c (hunk 26 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/alter_table.c:498]: 1ec3c304956b28fb39ffe63a72abc413f4383e72

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c (hunk 27 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:291]: 1ec3c304956b28fb39ffe63a72abc413f4383e72

File: src/backend/distributed/commands/utility_hook.c (hunk 20 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c]: 2. -> 0c86fc11a40e8845dc37d95f2343871f15d20d60

File: src/test/regress/sql/auto_undist_citus_local.sql (hunk 22 above)
Comment (onurctirtir) [reply src/test/regress/sql/auto_undist_citus_local.sql:317]: added those in commits 7b3deca & d77bfc9

File: src/backend/distributed/commands/create_citus_local_table.c [hunk 28]
```diff
@@ -80,6 +80,22 @@ create_citus_local_table(PG_FUNCTION_ARGS)
...
+		 * enabled, we automatically undistribute citus local tables that are
+		 * not chained with any reference tables back to postgres tables.
+		 * So give a warning to user for that.
+		 */
+		ereport(WARNING, (errmsg("citus local tables that are not chained with "
+								 "reference tables via foreign keys might be "
+								 "automatically converted back to postgres tables"),
+						  errhint("Consider setting citus.enable_local_reference_table_foreign_keys "
+								  "to 'off' to disable automatically undistributing "
+								  "citus local tables")))
```
Comment (onurctirtir) [inline src/backend/distributed/commands/create_citus_local_table.c:96]: is disabling the guc a good hint ?

**Review by marcocitus — APPROVED** [review]
A transaction block with 2+ operations (e.g. drop table and drop constraint) could be good.

File: src/backend/distributed/commands/utility_hook.c [hunk 29]
```diff
@@ -647,6 +687,138 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+		return false;
+	}
+
+	if (!InCoordinatedTransaction())
+	{
+		/* not interacting with any Citus objects */
+		return false;
+	}
+
+	if (IsCitusInitiatedBackend())
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:774]: Hmm, we use the concept `CitusInitiatedBackend` in other places as well, such as `MarkCitusInitiatedCoordinatorBackend()`.

To keep things less complicated, we should probably rename this to something else, such as `IsCitusInitiatedRemoteBackend ` or such

File: src/test/regress/expected/auto_undist_citus_local.out [hunk 30]
```diff
@@ -0,0 +1,635 @@
...
+LANGUAGE C STRICT
+AS 'citus', $$drop_constraint_cascade_via_perform_deletion$$;
+BEGIN;
+  SELECT drop_constraint_cascade_via_perform_deletion('citus_local_table', 'fkey_local_to_ref');
+ drop_constraint_cascade_via_perform_deletion
+---------------------------------------------------------------------
+
+(1 row)
+
+  -- we dropped constraint without going through utility hook,
```
Comment (onderkalaci) [inline src/test/regress/expected/auto_undist_citus_local.out:48]: Well, the behavior is unexpected :/  The constraint is dropped but Citus still considers the relationship as if the fkey exists. I guess this is a side-effect of our implementation.  We should probably reflect this in this comment?

Comment (onurctirtir) [reply src/test/regress/expected/auto_undist_citus_local.out:48]: yeah I mentioned this in https://github.com/citusdata/citus/pull/4538/files#diff-332a5456e294758640f5f05311fad43657f94d7d265d20f2e08aacc104ab0efdR166-R172

should this be mentioned in test comment too ?

File: src/backend/distributed/commands/utility_hook.c (hunk 29 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:774]: makes sense, I can open a seperate pr for that

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:774]: https://github.com/citusdata/citus/pull/4562

File: src/backend/distributed/commands/utility_hook.c [hunk 31]
```diff
@@ -237,8 +244,41 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+				UndistributeDisconnectedCitusLocalTables();
+			}
+			ResetConstraintDropped();
+		}
+
+		UtilityHookLevel--;
+	}
+	PG_CATCH();
+	{
+		if (UtilityHookLevel == 1)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:272]: I was initially thinking that we'd reset the constraint irrespective of `UtilityHookLevel`, but I guess they are equivalent anyway as we keep PG_RE_THROW. In fact, this is nicer as we call it exactly where we need to.
